File: src/main/java/top/jiangliuhong/wos/sql/parser/Parsed.java
/*
 * Copyright 2020 top.jiangliuhong.wos.sql
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package top.jiangliuhong.wos.sql.parser;

import lombok.Getter;
import lombok.Setter;
import top.jiangliuhong.wos.sql.def.Field;
import top.jiangliuhong.wos.sql.def.Table;
import top.jiangliuhong.wos.sql.exception.ParsingException;
import top.jiangliuhong.wos.sql.util.SqlStringUtil;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


@Getter
@Setter
public final class Parsed {


    private Table table;
    private String keyName;
    private Field keyField;
    private Map<String, Field> fieldMap = new HashMap<>();

    private final List<Field> tagFieldList = new ArrayList<>();

    private List<BeanElement> beanElementList;

    private Map<String, BeanElement> elementMap = new HashMap<>();
    private Map<String, String> propertyMapperMap = new HashMap<>();
    private Map<String, String> mapperPropertyMapLower = new HashMap<>();

    public Parsed(Table table) {
        this.table = table;
    }

    public Field getField(String name) {
        return fieldMap.get(name);
    }

    public Field getFieldExisted(String name) {
        Field field = fieldMap.get(name);
        if (field == null) {
            String message = String.format("Not exist %s from %s", name, this.table.getName());
            throw new ParsingException(message);
        }
        return field;
    }


    public String getTableName(String alia) {
        if (SqlStringUtil.isNullOrEmpty(alia))
            return this.getTableName();
        if (!alia.equalsIgnoreCase(getTableName()))
            return alia;
        return this.getTableName();
    }

    public String getTableName() {
        return this.table.getName();
    }


    public String getMapper(String property) {
        return propertyMapperMap.get(property);
    }


}

File: src/main/kotlin/top/jiangliuhong/wos/sql/parser/Parsed.kt
package top.jiangliuhong.wos.sql.parser

import java.util.*
import java.lang.reflect.Field;
import java.sql.Timestamp;

class Parsed {

    private var clzz: Class<*>? = null
    private val tableName: String? = null
    private var originTable: String? = null
    private val isNoSpec = true

    private var key: String? = null
    private var keyField: Field? = null
    private var tagKeyField: Field? = null
    private val tagFieldList: List<Field> = ArrayList<Field>()

    private var beanElementList: List<BeanElement>? = null

    private val elementMap: MutableMap<String, BeanElement> = HashMap<String, BeanElement>()
    private val propertyMapperMap: MutableMap<String, String> = HashMap()
    private val mapperPropertyMapLower: MutableMap<String, String> = HashMap()

    private val isNoCache = false

    fun getClzz(): Class<*>? {
        return clzz
    }

    fun setClzz(clzz: Class<*>?) {
        this.clzz = clzz
    }

    fun Parsed(clzz: Class<*>?) {
        this.clzz = clzz
    }


    fun getElement(property: String): BeanElement? {
        return elementMap[property]
    }

    fun getElementExisted(property: String): BeanElement {
        val be: BeanElement = elementMap[property] ?: throw ParsingException(
            "Not exist: "
                    + property
        )
        return be
    }

    fun getKey(): String? {
        return key
    }

    fun setKeyField(keyField: Field) {
        this.keyField = keyField
        this.key = keyField.getName()
    }

    fun getKeyField(): Field? {
        return keyField
    }

    fun getTagKeyField(): Field? {
        return tagKeyField
    }

    fun setTagKeyField(tagKeyField: Field?) {
        this.tagKeyField = tagKeyField
    }

    fun getTagFieldList(): List<Field> {
        return tagFieldList
    }

    fun isAutoIncreaseId(keyOneValue: Long?): Boolean {
        if (keyField == null) return false
        val keyOneType: Class<*> = keyField.getType()
        return (keyOneType != String::class.java && keyOneType != Date::class.java && keyOneType != Timestamp::class.java) && (keyOneValue == null || keyOneValue == 0L)
    }

    fun tryToGetLongKey(obj: Any?): Long? {
        var keyOneValue = 0L
        if (keyField == null) return null
        val keyOneType: Class<*> = keyField.getType()
        if (keyOneType != String::class.java && keyOneType != Date::class.java && keyOneType != Timestamp::class.java) {
            try {
                val keyValue: Any = keyField.get(obj)
                if (keyValue != null) {
                    keyOneValue = keyValue.toString().toLong()
                }
            } catch (e: Exception) {
                throw ParsingException(e.message)
            }
        }
        return keyOneValue
    }

    fun getBeanElementList(): List<BeanElement>? {
        return beanElementList
    }

    fun setOriginTable(originTable: String?) {
        this.originTable = originTable
    }

    fun reset(beanElementList: List<BeanElement>?) {
        this.beanElementList = beanElementList
        propertyMapperMap.clear()
        elementMap.clear()
        mapperPropertyMapLower.clear()
        for (e in this.beanElementList) {
            val property: String = e.getProperty()
            val mapper: String = e.getMapper()
            elementMap[property] = e
            propertyMapperMap[property] = mapper
            mapperPropertyMapLower[mapper.lowercase(Locale.getDefault())] = property
        }
    }


    fun getTableName(alia: String): String? {
        if (SqliStringUtil.isNullOrEmpty(alia)) return tableName
        if (!alia.equals(getClzName(), ignoreCase = true)) return alia
        return tableName
    }
}